### DISCUSSION TITLE
Provide success as a second argument to RequestOptions.deserialize

### DISCUSSION THREAD

**Discussion by romme** [MithrilJS/mithril.js#2924] [Ideas]
Long time user of the framework here. Actually it's the only framework i ever used, but nevermind, on to the feature request.

In case of REST apis which indicate failure or success as http status codes it may be cumbersome to tell error and success apart from the response object alone. Fortunately, Mithril already looks at the http status code and rejects the promise appropriately. But in case the promise is going to be rejected soon it may be undesireable to call deserialize on the response payload. What about adding a second argument to it so deserializing could be skipped?

```javascript
try {
    this.backend_response = await m.request({
        url,
        deserialize(response) {
            if (!(response instanceof Array)) // very crude attempt to tell a json payload from a json error object
                return response;
            for (var row of response) // response does not have an "error" field we could easily look for
                row.Date = new Date(t.Date);
            return response;
        },
    });
} catch (e) {
    alert(e.message); // in this case a json error object is a string
}
```

After implementing this feature:

```javascript
try {
    this.backend_response = await m.request({
        url,
        deserialize(response, success) {
            if (!success)
                return response;
            for (var row of response)
                row.Date = new Date(t.Date);
            return response;
        },
    });
} catch (e) {
    alert(e.message);
}
```

After writing all of this i realized that i could just deserialize manually after resolving the promise, which i'm going to do now. Still going to post this as an excuse to thank you for the great framework.

**Comment by osban** [comment]
@romme maybe you could deal with things using the `extract` method, something like [this](https://flems.io/#0=N4IgZglgNgpgziAXAbVAOwIYFsZJAOgAsAXLKEAGhAGMB7NYmBvEAXwvW10QICsEqdBk2J4hcYgAJGEyQF5JGOAE801SQAoAlPIB8k4AB00k6QCdlB46dPipZ+PMUB3DBClZ8DgI4BXeMQaAOQkxAAOcIgA9FFYygC0-PTxcDBmAG5p+MTKYRB0ACYw+HRYUTl5hTBRRVi0NQBGQRRWJjamMAAexGYY1MSImsASGMS+cC0OcGH0qaw6crrW7TYjY3CS+gAsAAw7yyuSAPzShGa0zpJTM2ipByuDAFIAygDyAHL4YRhmqRrXsxgWnu83u4losHwUFoAHN-vBgW1WMtqKNqIQNGkzDojG1TBhYGZAkEsbQzM1JFjEaZkWhacYZIEtJQaLQsHlCXgGhgGjByFRUrB+hBZngAIxbRA7NgcECYHB4EpwASs4TMHhsAC6VCgEDQAGsEChOAqeFh3GdoCzfGZyDxQhFolFfGgwvqYSU2bELWZoAABABM+B2+C23uIlqg+HNaHw-BZFW4IDg1F9YVErE1rCAA)?

**Comment by romme** [comment]
@osban that's rougly the same amount of code, went with [this](https://flems.io/#0=N4IgZglgNgpgziAXAbVAOwIYFsZJAOgAsAXLKEAGhAGMB7NYmBvEAXwvW10QICsEqdBk2J4hcYgAJGEyQF5JGOAE801SQAoAlPIB8k4AB00k6QCdlB46dPipZ+PMUB3DBClZ8DgI4BXeMQaAOQkxAAOcIgA9FFYygC0-PTxcDBmAG5p+MTKYRB0ACYw+HRYUTl5hTBRRVi0NQBGQVrWNmC0ZprpGJ1htLK0YJIOcPh9EnA648T4EAWSAFQKAIwADKsA3JIxkkWpZhAYUBAAXjCttvRwtLD4ULQA5hojLSaSrK3UGMTUhBppZh0RjepiOaUCQQBHSCFEkANepg+aCRxhkgS0lBotCweVgZjwDQwDRg5CoqVg1GIECueAAzIhVmwOCBMDg8CU4AIscJmDw2ABdKjHNAAawQKE4bJ4WHchAOpJAvjM5B4oQi0SivjQYRFDxK2NisvlAAEAEz4Vb4AAshuIcug+BlaHw-ExFW4IDg1AOYVErH5rCAA) in the end

**Comment by osban** [comment]
@romme Ah cool. Yeah I wasn't quite sure how much control you needed, so if this works, it's the best option 🙂
